### ISSUE TITLE
2.0.10 ActiveForm breaks form submission in certain instances

### ISSUE THREAD

**Issue by cyphix333** [yiisoft/yii2#12803]
I recently upgraded to 2.0.10 and I noticed some weird behaviour with my form submissions; my forms would no longer submit unless I was submitting them via an AJAX procedure.

So I did some debugging, on all my _non-ajax_ forms I have this binded to them...

```
$('form').not('.js-ajax').on('submit', function(e) {
    // Disable submit button
    $(e.target).find('button[type=submit], input[type=submit]').prop('disabled', true);
});
```

So as you can probably tell it disables the submit button when the user submits the form; if I comment out this code then the form will _now_ **submit**; but the forms were submitting fine on 2.0.9 with this code enabled, which I confirmed by going back and forth between 2.0.9 and 2.0.10.

Now you ask what does this have to do with ActiveForm? Well this only happens on forms made with ActiveForm - I have a _non-ajax_ form done _without_ ActiveForm and it submits fine.

So, a summary:
## 2.0.10

Form submitted via AJAX: **works**
Form made with ActiveForm _with_ submit button disabling: **Doesn't work**
Form made with ActiveForm _without_ submit button disabling: **Works**
Form NOT made with ActiveForm: **works**
## 2.0.9 and below

**All situations work**
### What steps will reproduce the problem?

Create standard and ActiveForm forms to test - test ActiveForm forms with and without submit button disabling on submit.
### What is the expected result?

Form sbumits.
### What do you get instead?

Forms you build with ActiveForm AND disable the submit button on submit don't submit.
### Additional info

| Q | A |
| --- | --- |
| Yii version | 2.0.10 |
| PHP version | 5.5.28 |
| Operating system | Win 7 |

**Comment by samdark** [comment]
Please post the code for opening the form and for submit button.

**Comment by cyphix333** [comment]
**Opening form:**

`<?php $form = ActiveForm::begin(); ?>`

**Submit button:**

`<?=Html::button(Yii::t('general', 'Login'), ['type' => 'submit', 'id' => 'login_submit', 'class' => 'primary'])?>`

**Comment by Thoulah** [comment]
Small sidenote: there is a `Html::submitButton`

**Comment by cyphix333** [comment]
@Thoulah Yeah, I'm using `Html::submitInput` on another form and it doesn't make any difference; but generally I prefer to use `button`'s - the `Html::submitInput` is probably a remnant from earlier code.

**Comment by dynasource** [comment]
@cyphix333, is your objective to prevent multiple submissions?

**Comment by samdark** [comment]
@dynasource yes. He's disabling the button after 1st submit.

**Comment by dynasource** [comment]
Ive thought about this one a while. In the past Ive been using events like this. Adding 'disabled' as attribute + adding CSS, but although it looks like its working, you can be misleaded. 

Different events for the submission of forms can override your events which makes your double submission prevention useless.

A trick which works 100% of the time though, without being 'disturbed' by other events, is the use of a ghost button. When you submit, a hidden 'brother' button is shown/created without functionality. You can click what you want on this brother, but it can never lead to a submission or trigger a submission event. Based on your configuration, you can show this brother for a few seconds.

**Comment by dynasource** [comment]
https://github.com/yiisoft/yii2/issues/10498#issuecomment-253542255

**Comment by cyphix333** [comment]
@dynasource Hmmmmm...... interesting technique; and yes, I can see it's merit. That's something I am definitely going to be looking at moving forward. :)

**Comment by dynasource** [comment]
Ill open a new topic for this

**Comment by samdark** [comment]
http://pastebin.com/91cUfrnv

**Comment by cyphix333** [comment]
@samdark Looks cool. Gonna have to give that a shot too, though I generally don't re-enable the submit button unless it's an ajax form (which I do via jQuerys _always_ promise method) as there isn't much point as it will be enabled again on the next page if there was an error.

**Comment by samdark** [comment]
Solution is by @SilverFire. I haven't tested it myself...

**Comment by makroxyz** [comment]
I have a similar issue with 2.0.10...
While 2.0.9 works fine 2.0.10 doesn't submit the form

```php
<?php $form = ActiveForm::begin([
        'id' => 'form-barcode',
        'options' => ['data-pjax' => '1'],
        'enableClientValidation' => false
 ]); ?>

....
....

<?php
$this->registerJs("
    $('#form-barcode').on('change', ':text', function() {
        $('#form-barcode').submit();
    });
    $(document).on('pjax:end', function() {
        $('#form-barcode :text').val('');
    });
");
```

**Comment by cyphix333** [comment]
@makroxyz Your post reminded me to mention that I am extending from the `ActiveForm` class, basically to set some parameters so I don't have to do it each time, so I thought I would mention what I am doing here:

This is all I'm changing:

**ActiveForm:**

`public $fieldClass = 'common\widgets\ActiveField';`

```
public $fieldConfig = [
    'template' => "{label}\n{beginWrapper}\n{input}\n{hint}\n{endWrapper}",
];
```
`public $enableClientValidation = false;`

**ActiveField:**

`public $checkboxTemplate = "<div class=\"checkbox\">\n{beginLabel}\n{input}\n{labelTitle}\n{endLabel}\n{hint}\n</div>";`

`public $radioTemplate = "<div class=\"radio\">\n{beginLabel}\n{input}\n{labelTitle}\n{endLabel}\n{hint}\n</div>";`

**Comment by PsychicCat** [comment]
thank god i was going nuts trying to figure out why my form wouldnt submit

**Comment by buddha87** [comment]
We had a similar issue with our loader button. 
Our submit buttons have a custom `click` handler attached, for rendering a loader animation and prevent multiple clicks.

The problem was introduced by the following line:
https://github.com/yiisoft/yii2/commit/3ba72da6fa11c7565de7ceb002bdf6ff86c841d4#diff-3096f348a33d8aa1d79ebe62e5e0f304R623

Since the form submission is now triggered by another `data.submitObject.trigger("click");` instead of calling `$form.submit();`, the handler is called twice, which was prevented by returing false before the workaround.

As a workaround, now we just block click events where 'evt.originalEvent' is present: https://github.com/humhub/humhub/blob/master/js/app.js#L176

In your case the handler is probably not triggered, since the click event has no effect on your disabled button.

**Comment by DmLapin** [comment]
Oh, one feature, one bug. We also has also encountered with this issue. As @buddha87 wrote, [this change](https://github.com/yiisoft/yii2/commit/3ba72da6fa11c7565de7ceb002bdf6ff86c841d4) brings problem.

@samdark, @VirtualRJ,  can you please clarify the reason of this change? It seems, that @AnatolyRugalev [has already made](https://github.com/yiisoft/yii2/commit/fa5baf32756a8193a380a8cf43251983bb16b221) support of HTML5 attributes in May'16 in more compatible way. In our application we also disable button to prevent any possible iteractions with it.

**Comment by VirtualRJ** [comment]
I agree it is in a more compatible way to the old way and specifically Yii, but taking the button/input options introduced in HTML5 and applying and restoring these to the form seems to me like a bit of a hack. No need for HTML5 to introduce these as it would also be possible to implement them purely on the basis of JS. 

The 'solution' I created was in response to a relatively old issue #8197 and I did fail to see there was another solution implementation. I still think my solution respects the HTML5 standard more and is in that way cleaner, but as we now notice it can break submission with different double submission prevention techniques. Therefore I also very much support the initiative of a Yii2 OOTB double submit prevention system.
